{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a99240b3",
   "metadata": {},
   "source": [
    "## Reproduce Figure 5b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "db325144",
   "metadata": {},
   "outputs": [],
   "source": [
    "import stash as st\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1870f70",
   "metadata": {},
   "source": [
    "## Launch predictions using Variantformer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0453ccd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Essential imports\n",
    "import sys\n",
    "from pathlib import Path\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "sys.path.append(str(Path.cwd().parent))\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "\n",
    "from processors.variantprocessor import VariantProcessor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96b29a32",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "37450a7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tissue\n",
       "blood                              17724\n",
       "skin                               14723\n",
       "adipose                            14671\n",
       "brain - frontal cortex (ba9)       13029\n",
       "brain - putamen (basal ganglia)    12332\n",
       "brain - substantia nigra           10604\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants = st.get('b23c9b69')\n",
    "variants['tissue'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28dbccb3",
   "metadata": {},
   "source": [
    "### Predict for first 10 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7d6c4f98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tiisue map to align tissue names between VF and Eqtl catalog data\n",
    "vf_tissue_map = {'skin' : 'skin - sun exposed (lower leg)',\n",
    "                 'blood' : 'whole blood',\n",
    "                 'adipose' : 'adipose - subcutaneous',\n",
    "                 'brain - frontal cortex (ba9)': 'brain - frontal cortex (ba9)',\n",
    "                 'brain - putamen (basal ganglia)': 'brain - putamen (basal ganglia)',\n",
    "                 'brain - substantia nigra' : 'brain - substantia nigra'\n",
    "                 }\n",
    "variants['tissue'] = variants['tissue'].map(vf_tissue_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "71fc07af",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "first10_variants = variants.groupby('tissue').sample(n=1, random_state=42)\n",
    "first10_variants['chr'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[0])\n",
    "first10_variants['pos'] = first10_variants['variant_id'].apply(lambda x: int(x.split('_')[1]))\n",
    "first10_variants['ref'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[2])\n",
    "first10_variants['alt'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[3])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "7ca44c3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tissue\n",
       "Whole_Blood                    17724\n",
       "Skin_Sun_Exposed_Lower_leg     14723\n",
       "Adipose_Subcutaneous           14671\n",
       "Brain_Frontal_Cortex_BA9       13029\n",
       "Brain_Putamen_basal_ganglia    12332\n",
       "Brain_Substantia_nigra         10604\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants['tissue'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "73844833",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-03 14:41:52 - processors.variantprocessor - INFO - Initializing Variant Processor...\n",
      "2025-11-03 14:41:52 - processors.multi_datasets_loader - INFO - Loading gene annotations...\n",
      "2025-11-03 14:41:52 - processors.multi_datasets_loader - INFO - Loading CRE annotations...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ðŸš€ Initializing VariantFormer Variant Processor...\n",
      "ðŸ”¬ Running VariantFormer variant analysis...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-03 14:41:53 - processors.variantprocessor - INFO - Loading variants...\n",
      "2025-11-03 14:41:53 - processors.multi_datasets_loader - INFO - Loaded 10 variants\n",
      "2025-11-03 14:41:53 - processors.variantprocessor - INFO - Loaded 10 variants for processing\n",
      "2025-11-03 14:41:53 - processors.variantprocessor - INFO - Mapped 10 gene-variant pairs\n",
      "2025-11-03 14:41:53 - processors.variantprocessor - INFO - Loading BPE encoder...\n",
      "2025-11-03 14:41:53 - processors.variantprocessor - INFO - Loading model...\n",
      "2025-11-03 14:41:53 - processors.model_manager - INFO - Loading Seq2Reg model...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded BPE vocabulary from /work/vocabs/bpe_vocabulary_500_using_huggingface.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-03 14:41:54 - processors.model_manager - INFO - Loading Seq2Reg gene model...\n",
      "2025-11-03 14:41:54 - processors.model_manager - INFO - Creating Seq2Gene model...\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO - Model class: <class 'seq2gene.model_combined_modulator.Seq2GenePredictorCombinedModulator'>\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO - Model architecture:\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO - Model: Seq2GenePredictorCombinedModulator\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO -   start_tkn: 96,768 params\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO -   cre_tokenizer: 31,826,153 params\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO -   gene_tokenizer: 31,826,153 params\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO -   gene_map: 787,968 params\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO -   cre_map: 787,968 params\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO -   combined_modulator: 1,157,298,176 params\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO -   tissue_heads: 4,726,273 params\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO -   gene_loss: 0 params\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO - Total number of parameters: 1,227,349,459\n",
      "2025-11-03 14:41:59 - processors.model_manager - INFO - Loading checkpoint from /work/_artifacts/v4_ag_epoch9_checkpoint.pth\n",
      "2025-11-03 14:42:04 - processors.model_manager - INFO - Model loaded successfully on cuda\n",
      "2025-11-03 14:42:04 - processors.variantprocessor - INFO - Initializing trainer...\n",
      "INFO: Using bfloat16 Automatic Mixed Precision (AMP)\n",
      "2025-11-03 14:42:04 - lightning.pytorch.utilities.rank_zero - INFO - Using bfloat16 Automatic Mixed Precision (AMP)\n",
      "INFO: GPU available: True (cuda), used: True\n",
      "2025-11-03 14:42:04 - lightning.pytorch.utilities.rank_zero - INFO - GPU available: True (cuda), used: True\n",
      "INFO: TPU available: False, using: 0 TPU cores\n",
      "2025-11-03 14:42:04 - lightning.pytorch.utilities.rank_zero - INFO - TPU available: False, using: 0 TPU cores\n",
      "INFO: HPU available: False, using: 0 HPUs\n",
      "2025-11-03 14:42:04 - lightning.pytorch.utilities.rank_zero - INFO - HPU available: False, using: 0 HPUs\n",
      "2025-11-03 14:42:04 - processors.variantprocessor - INFO - Initialization complete!\n",
      "INFO: LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n",
      "2025-11-03 14:42:04 - lightning.pytorch.accelerators.cuda - INFO - LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/cres_seqs_manifest.parquet\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3e0e45a13ae940c085adf61de22e9ff7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Predicting: |          | 0/? [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Loading parquet file: /tmp/tmpcgcsqwid/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Loading parquet file: /tmp/tmpcgcsqwid/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Loading parquet file: /tmp/tmpcgcsqwid/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Loading parquet file: /tmp/tmpcgcsqwid/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Loading parquet file: /tmp/tmpcgcsqwid/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Loading parquet file: /tmp/tmpcgcsqwid/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Loading parquet file: /tmp/tmpcgcsqwid/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Validated schema - found columns: {'chromosome', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Validated schema - found columns: {'chromosome', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Validated schema - found columns: {'chromosome', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Validated schema - found columns: {'chromosome', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Validated schema - found columns: {'chromosome', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Validated schema - found columns: {'chromosome', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Validated schema - found columns: {'chromosome', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data_split/eas/cres/data_split/HG00404_chr10.pkl.gz\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data_split/eas/cres/data_split/HG00404_chr10.pkl.gz\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data_split/afr/cres/data_split/HG01879_chr10.pkl.gz\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data_split/eur/cres/data_split/HG00096_chr10.pkl.gz\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data_split/hg38/cres/data_split/hg38_chr10.pkl.gz\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data_split/amr/cres/data_split/HG00551_chr10.pkl.gz\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data_split/sas/cres/data_split/HG01589_chr10.pkl.gz\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Loading parquet file: /tmp/tmpcgcsqwid/model/common/reference_genomes/cres_seqs_manifest.parquet\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Validated schema - found columns: {'chromosome', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:05 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data_split/hg38/cres/data_split/hg38_chr10.pkl.gz\n",
      "2025-11-03 14:42:07 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:07 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:07 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:07 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:07 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:07 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:07 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:07 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Loading parquet file: /tmp/tmpsjlicydj/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path'}\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Loading parquet file: /tmp/tmpsjlicydj/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Loading parquet file: /tmp/tmpsjlicydj/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Loading parquet file: /tmp/tmpsjlicydj/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Loading parquet file: /tmp/tmpsjlicydj/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Loading parquet file: /tmp/tmpsjlicydj/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes/ENSG00000095485.16/gene_vocab.csv\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path'}\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Loading parquet file: /tmp/tmpsjlicydj/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path'}\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path'}\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path'}\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path'}\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Loading parquet file: /tmp/tmpsjlicydj/model/common/cres_all_genes_manifest.parquet\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path'}\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes/ENSG00000095485.16/gene_vocab.csv\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path'}\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes/ENSG00000095485.16/gene_vocab.csv\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes/ENSG00000095485.16/gene_vocab.csv\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes/ENSG00000095485.16/gene_vocab.csv\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes/ENSG00000095485.16/gene_vocab.csv\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes/ENSG00000095485.16/gene_vocab.csv\n",
      "2025-11-03 14:42:08 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/cres_all_genes/ENSG00000095485.16/gene_vocab.csv\n",
      "2025-11-03 14:42:10 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:10 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:10 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:10 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:10 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:10 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:10 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:10 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Loading parquet file: /tmp/tmpeu6wg81_/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Loading parquet file: /tmp/tmpeu6wg81_/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Loading parquet file: /tmp/tmpeu6wg81_/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Loading parquet file: /tmp/tmpeu6wg81_/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Loading parquet file: /tmp/tmpeu6wg81_/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Loading parquet file: /tmp/tmpeu6wg81_/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Loading parquet file: /tmp/tmpeu6wg81_/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Loading parquet file: /tmp/tmpeu6wg81_/model/common/reference_genomes/genes_seqs_manifest.parquet\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Validated schema - found columns: {'gene_id', 'file_path', 'population', '__index_level_0__'}\n",
      "2025-11-03 14:42:11 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data/eur/genes/data/ENSG00000095485.16_HG00096.npz\n",
      "2025-11-03 14:42:12 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data/hg38/genes/data/ENSG00000095485.16_hg38.npz\n",
      "2025-11-03 14:42:12 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data/afr/genes/data/ENSG00000095485.16_HG01879.npz\n",
      "2025-11-03 14:42:12 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data/amr/genes/data/ENSG00000095485.16_HG00551.npz\n",
      "2025-11-03 14:42:12 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data/sas/genes/data/ENSG00000095485.16_HG01589.npz\n",
      "2025-11-03 14:42:12 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data/eas/genes/data/ENSG00000095485.16_HG00404.npz\n",
      "2025-11-03 14:42:12 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data/hg38/genes/data/ENSG00000095485.16_hg38.npz\n",
      "2025-11-03 14:42:12 - utils.assets - INFO - Downloading from S3: s3://czi-variantformer/model/common/reference_genomes/data/eas/genes/data/ENSG00000095485.16_HG00404.npz\n",
      "2025-11-03 14:42:41 - processors.variantprocessor - INFO - Predictions saved to /tmp/vep_multi_output/vep_VF.parquet\n",
      "2025-11-03 14:42:41 - processors.variantprocessor - INFO - Cleanup complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formatting VariantFormer scores and computing eQTL statistics...\n",
      "Computing eQTL statistics...\n"
     ]
    }
   ],
   "source": [
    "# Initialize VariantFormer\n",
    "print(\"ðŸš€ Initializing VariantFormer Variant Processor...\")\n",
    "model_class = 'v4_ag' # model class can be 'v4_ag', 'v4_pcg'. AG model is all-genes model trained on both protein-coding and non-coding genes.\n",
    "# model_class = 'v4_pcg' # Uncomment to use the PCG model\n",
    "vep = VariantProcessor(model_class=model_class)\n",
    "# Run variant predictions\n",
    "print(\"ðŸ”¬ Running VariantFormer variant analysis...\")\n",
    "output_dir = \"/tmp/vep_multi_output\"\n",
    "# Predict expression effects for all variants\n",
    "raw_predictions = vep.predict(var_df=first10_variants, output_dir=output_dir)\n",
    "print(\"Formatting VariantFormer scores and computing eQTL statistics...\")\n",
    "formatted_scores = vep.format_scores(raw_predictions)\n",
    "print(\"Computing eQTL statistics...\")\n",
    "final_results = vep.eqtl_scores(formatted_scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6c2351f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>genes</th>\n",
       "      <th>tissues</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>VF-agg-log2fc-weighted</th>\n",
       "      <th>VF-AFR-2-exp-log2fc</th>\n",
       "      <th>VF-AMR-2-exp-log2fc</th>\n",
       "      <th>VF-EAS-2-exp-log2fc</th>\n",
       "      <th>VF-EUR-2-exp-log2fc</th>\n",
       "      <th>VF-REF_HG38-2-exp-log2fc</th>\n",
       "      <th>VF-SAS-2-exp-log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10_100236603_C_T</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>adipose - subcutaneous</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100236603</td>\n",
       "      <td>0.009466</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027943</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007029</td>\n",
       "      <td>0.007029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_100236603_C_T</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100236603</td>\n",
       "      <td>0.017387</td>\n",
       "      <td>0.060486</td>\n",
       "      <td>0.006830</td>\n",
       "      <td>0.020407</td>\n",
       "      <td>0.020407</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_100238657_T_C</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>adipose - subcutaneous</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100238657</td>\n",
       "      <td>0.009475</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027943</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007029</td>\n",
       "      <td>0.007029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10_100238657_T_C</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100238657</td>\n",
       "      <td>0.017376</td>\n",
       "      <td>0.060486</td>\n",
       "      <td>0.006830</td>\n",
       "      <td>0.020407</td>\n",
       "      <td>0.020407</td>\n",
       "      <td>0.006830</td>\n",
       "      <td>0.020407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10_100239989_G_A</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>adipose - subcutaneous</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100239989</td>\n",
       "      <td>0.009435</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027943</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007029</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            variant_id               genes                 tissues ref alt  \\\n",
       "0  chr10_100236603_C_T  ENSG00000095485.16  adipose - subcutaneous   C   T   \n",
       "1  chr10_100236603_C_T  ENSG00000095485.16             whole blood   C   T   \n",
       "2  chr10_100238657_T_C  ENSG00000095485.16  adipose - subcutaneous   T   C   \n",
       "3  chr10_100238657_T_C  ENSG00000095485.16             whole blood   T   C   \n",
       "4  chr10_100239989_G_A  ENSG00000095485.16  adipose - subcutaneous   G   A   \n",
       "\n",
       "     chr        pos  VF-agg-log2fc-weighted  VF-AFR-2-exp-log2fc  \\\n",
       "0  chr10  100236603                0.009466             0.000000   \n",
       "1  chr10  100236603                0.017387             0.060486   \n",
       "2  chr10  100238657                0.009475             0.000000   \n",
       "3  chr10  100238657                0.017376             0.060486   \n",
       "4  chr10  100239989                0.009435             0.000000   \n",
       "\n",
       "   VF-AMR-2-exp-log2fc  VF-EAS-2-exp-log2fc  VF-EUR-2-exp-log2fc  \\\n",
       "0             0.027943             0.000000             0.000000   \n",
       "1             0.006830             0.020407             0.020407   \n",
       "2             0.027943             0.000000             0.000000   \n",
       "3             0.006830             0.020407             0.020407   \n",
       "4             0.027943             0.000000             0.000000   \n",
       "\n",
       "   VF-REF_HG38-2-exp-log2fc  VF-SAS-2-exp-log2fc  \n",
       "0                  0.007029             0.007029  \n",
       "1                  0.000000             0.020407  \n",
       "2                  0.007029             0.007029  \n",
       "3                  0.006830             0.020407  \n",
       "4                  0.000000             0.007029  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "677df43c",
   "metadata": {},
   "source": [
    "## Launch predictions using Alphagenome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "747b1f54",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "489154a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = st.get('b23c9b69')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "99e32323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tiisue map to align tissue names between VF and Eqtl catalog data\n",
    "alphagenome_tissue_map = {'skin' : 'Skin_Sun_Exposed_Lower_leg',\n",
    "                 'blood' : 'Whole_Blood',\n",
    "                 'adipose' : 'Adipose_Subcutaneous',\n",
    "                 'brain - frontal cortex (ba9)': 'Brain_Frontal_Cortex_BA9',\n",
    "                 'brain - putamen (basal ganglia)': 'Brain_Putamen_basal_ganglia',\n",
    "                 'brain - substantia nigra' : 'Brain_Substantia_nigra'\n",
    "                 }\n",
    "variants['tissue'] = variants['tissue'].map(alphagenome_tissue_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64b7e620",
   "metadata": {},
   "outputs": [],
   "source": [
    "first10_variants = variants.groupby('tissue').sample(n=10, random_state=42)\n",
    "first10_variants['CHROM'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[0])\n",
    "first10_variants['POS'] = first10_variants['variant_id'].apply(lambda x: int(x.split('_')[1]))\n",
    "first10_variants['REF'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[2])\n",
    "first10_variants['ALT'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[3])\n",
    "first10_variants['gene_id'] = first10_variants['gene_id'].apply(lambda x: x.split('.')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "32b03eba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2mUsing Python 3.12.10 environment at: /work/.venv\u001b[0m\n",
      "\u001b[2mAudited \u001b[1m1 package\u001b[0m \u001b[2min 8ms\u001b[0m\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "# Install alphagenome\n",
    "!uv pip install alphagenome\n",
    "from io import StringIO\n",
    "from alphagenome import colab_utils\n",
    "from alphagenome.data import genome\n",
    "from alphagenome.models import dna_client, variant_scorers\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf62518b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the model.\n",
    "API_KEY = \"your_api_key_here\"\n",
    "dna_model = dna_client.create(API_KEY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e10d15e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alphagenome specifications\n",
    "organism = 'human'\n",
    "# @markdown Specify length of sequence around variants to predict:\n",
    "sequence_length = '1MB'\n",
    "sequence_length = dna_client.SUPPORTED_SEQUENCE_LENGTHS[\n",
    "    f'SEQUENCE_LENGTH_{sequence_length}'\n",
    "]\n",
    "score_rna_seq = True\n",
    "download_predictions = False\n",
    "# Parse organism specification.\n",
    "organism_map = {\n",
    "    'human': dna_client.Organism.HOMO_SAPIENS,\n",
    "    'mouse': dna_client.Organism.MUS_MUSCULUS,\n",
    "}\n",
    "organism = organism_map[organism]\n",
    "# Parse scorer specification.\n",
    "scorer_selections = {\n",
    "    'rna_seq': score_rna_seq,\n",
    "}\n",
    "all_scorers = variant_scorers.RECOMMENDED_VARIANT_SCORERS\n",
    "selected_scorers = [\n",
    "    all_scorers[key]\n",
    "    for key in all_scorers\n",
    "    if scorer_selections.get(key.lower(), False)\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "87bae8b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "for i, vcf_row in first10_variants.iterrows():\n",
    "  variant = genome.Variant(\n",
    "      chromosome=str(vcf_row.CHROM),\n",
    "      position=int(vcf_row.POS),\n",
    "      reference_bases=vcf_row.REF,\n",
    "      alternate_bases=vcf_row.ALT,\n",
    "      name=vcf_row.variant_id,\n",
    "  )\n",
    "  interval = variant.reference_interval.resize(sequence_length)\n",
    "\n",
    "  variant_scores = dna_model.score_variant(\n",
    "      interval=interval,\n",
    "      variant=variant,\n",
    "      variant_scorers=selected_scorers,\n",
    "      organism=organism,\n",
    "  )\n",
    "  df_scores = variant_scorers.tidy_scores([variant_scores])\n",
    "  tissue_key = vcf_row.tissue\n",
    "  df = df_scores[(df_scores['gene_id'] == vcf_row.gene_id) & (df_scores['gtex_tissue'] == tissue_key)]['raw_score']\n",
    "  vcf_row['alphagenome_score_new'] = df.values[0].mean() if not df.empty else np.nan\n",
    "  results.append(vcf_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d4afb797",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82341c60",
   "metadata": {},
   "source": [
    "**The alphagenome score might deviate a little because we believe the underlying model is evolving** "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2029fe2f",
   "metadata": {},
   "source": [
    "## Visualize all the precalculated scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "747f1f32",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = st.get('b23c9b69')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e6ae6fba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>slope</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>maf</th>\n",
       "      <th>VF-agg-log2fc-weighted_ag</th>\n",
       "      <th>VF-AFR-2-log2fc_ag</th>\n",
       "      <th>...</th>\n",
       "      <th>VF-SAS-2-log2fc_ag</th>\n",
       "      <th>VF-agg-log2fc-weighted_pcg</th>\n",
       "      <th>VF-AFR-2-log2fc_pcg</th>\n",
       "      <th>VF-AMR-2-log2fc_pcg</th>\n",
       "      <th>VF-EAS-2-log2fc_pcg</th>\n",
       "      <th>VF-EUR-2-log2fc_pcg</th>\n",
       "      <th>VF-SAS-2-log2fc_pcg</th>\n",
       "      <th>variant_id_gene</th>\n",
       "      <th>Borzoi-log2FC</th>\n",
       "      <th>alphagenome_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10_100236603_C_T</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>adipose</td>\n",
       "      <td>0.659835</td>\n",
       "      <td>2.606480e-38</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.429134</td>\n",
       "      <td>0.009466</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007029</td>\n",
       "      <td>0.012622</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041915</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>chr10_100236603_C_T_ENSG00000095485.16</td>\n",
       "      <td>-0.000599</td>\n",
       "      <td>0.000815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_100236603_C_T</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>blood</td>\n",
       "      <td>0.476792</td>\n",
       "      <td>8.194630e-18</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.410256</td>\n",
       "      <td>0.017387</td>\n",
       "      <td>0.060486</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020407</td>\n",
       "      <td>0.007350</td>\n",
       "      <td>0.033332</td>\n",
       "      <td>0.006720</td>\n",
       "      <td>0.020079</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.020079</td>\n",
       "      <td>chr10_100236603_C_T_ENSG00000095485.16</td>\n",
       "      <td>-0.000461</td>\n",
       "      <td>0.000996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_100238657_T_C</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>adipose</td>\n",
       "      <td>0.659753</td>\n",
       "      <td>2.629150e-38</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.429134</td>\n",
       "      <td>0.009475</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007029</td>\n",
       "      <td>0.013702</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041915</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.007058</td>\n",
       "      <td>chr10_100238657_T_C_ENSG00000095485.16</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>-0.000954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10_100238657_T_C</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>blood</td>\n",
       "      <td>0.476752</td>\n",
       "      <td>8.216580e-18</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.410256</td>\n",
       "      <td>0.017376</td>\n",
       "      <td>0.060486</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020407</td>\n",
       "      <td>0.011797</td>\n",
       "      <td>0.033332</td>\n",
       "      <td>0.006720</td>\n",
       "      <td>0.026719</td>\n",
       "      <td>0.00672</td>\n",
       "      <td>0.026719</td>\n",
       "      <td>chr10_100238657_T_C_ENSG00000095485.16</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>-0.000968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10_100239989_G_A</td>\n",
       "      <td>ENSG00000095485.16</td>\n",
       "      <td>adipose</td>\n",
       "      <td>0.659729</td>\n",
       "      <td>2.636980e-38</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.429134</td>\n",
       "      <td>0.009435</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007029</td>\n",
       "      <td>0.012604</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041915</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>chr10_100239989_G_A_ENSG00000095485.16</td>\n",
       "      <td>0.001238</td>\n",
       "      <td>0.000725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            variant_id             gene_id   tissue     slope        pvalue  \\\n",
       "0  chr10_100236603_C_T  ENSG00000095485.16  adipose  0.659835  2.606480e-38   \n",
       "1  chr10_100236603_C_T  ENSG00000095485.16    blood  0.476792  8.194630e-18   \n",
       "2  chr10_100238657_T_C  ENSG00000095485.16  adipose  0.659753  2.629150e-38   \n",
       "3  chr10_100238657_T_C  ENSG00000095485.16    blood  0.476752  8.216580e-18   \n",
       "4  chr10_100239989_G_A  ENSG00000095485.16  adipose  0.659729  2.636980e-38   \n",
       "\n",
       "  ref alt       maf  VF-agg-log2fc-weighted_ag  VF-AFR-2-log2fc_ag  ...  \\\n",
       "0   C   T  0.429134                   0.009466            0.000000  ...   \n",
       "1   C   T  0.410256                   0.017387            0.060486  ...   \n",
       "2   T   C  0.429134                   0.009475            0.000000  ...   \n",
       "3   T   C  0.410256                   0.017376            0.060486  ...   \n",
       "4   G   A  0.429134                   0.009435            0.000000  ...   \n",
       "\n",
       "   VF-SAS-2-log2fc_ag  VF-agg-log2fc-weighted_pcg  VF-AFR-2-log2fc_pcg  \\\n",
       "0            0.007029                    0.012622             0.000000   \n",
       "1            0.020407                    0.007350             0.033332   \n",
       "2            0.007029                    0.013702             0.000000   \n",
       "3            0.020407                    0.011797             0.033332   \n",
       "4            0.007029                    0.012604             0.000000   \n",
       "\n",
       "   VF-AMR-2-log2fc_pcg  VF-EAS-2-log2fc_pcg  VF-EUR-2-log2fc_pcg  \\\n",
       "0             0.041915             0.000000              0.00000   \n",
       "1             0.006720             0.020079              0.00000   \n",
       "2             0.041915             0.000000              0.00000   \n",
       "3             0.006720             0.026719              0.00672   \n",
       "4             0.041915             0.000000              0.00000   \n",
       "\n",
       "   VF-SAS-2-log2fc_pcg                         variant_id_gene  Borzoi-log2FC  \\\n",
       "0             0.000000  chr10_100236603_C_T_ENSG00000095485.16      -0.000599   \n",
       "1             0.020079  chr10_100236603_C_T_ENSG00000095485.16      -0.000461   \n",
       "2             0.007058  chr10_100238657_T_C_ENSG00000095485.16       0.000460   \n",
       "3             0.026719  chr10_100238657_T_C_ENSG00000095485.16       0.000305   \n",
       "4             0.000000  chr10_100239989_G_A_ENSG00000095485.16       0.001238   \n",
       "\n",
       "   alphagenome_score  \n",
       "0           0.000815  \n",
       "1           0.000996  \n",
       "2          -0.000954  \n",
       "3          -0.000968  \n",
       "4           0.000725  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d5442ac",
   "metadata": {},
   "source": [
    "### VF results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9b719f41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>spearman_correlation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>adipose</td>\n",
       "      <td>0.540437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>blood</td>\n",
       "      <td>0.453173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>brain - frontal cortex (ba9)</td>\n",
       "      <td>0.625611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brain - putamen (basal ganglia)</td>\n",
       "      <td>0.607318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brain - substantia nigra</td>\n",
       "      <td>0.587621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>skin</td>\n",
       "      <td>0.605895</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            tissue  spearman_correlation\n",
       "0                          adipose              0.540437\n",
       "1                            blood              0.453173\n",
       "2     brain - frontal cortex (ba9)              0.625611\n",
       "3  brain - putamen (basal ganglia)              0.607318\n",
       "4         brain - substantia nigra              0.587621\n",
       "5                             skin              0.605895"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Tissue specific spoearman correlation plots\n",
    "vf_spearman = variants.groupby('tissue').apply(\n",
    "    lambda x: x['slope'].corr(x['VF-agg-log2fc-weighted_ag'], method='spearman')\n",
    ").reset_index()\n",
    "vf_spearman.columns = ['tissue', 'spearman_correlation']\n",
    "vf_spearman\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77355fe9",
   "metadata": {},
   "source": [
    "### Alphagenome results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "44f0e516",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>spearman_correlation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>adipose</td>\n",
       "      <td>0.039658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>blood</td>\n",
       "      <td>0.035846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>brain - frontal cortex (ba9)</td>\n",
       "      <td>0.044307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brain - putamen (basal ganglia)</td>\n",
       "      <td>0.016789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brain - substantia nigra</td>\n",
       "      <td>0.018051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>skin</td>\n",
       "      <td>0.019730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            tissue  spearman_correlation\n",
       "0                          adipose              0.039658\n",
       "1                            blood              0.035846\n",
       "2     brain - frontal cortex (ba9)              0.044307\n",
       "3  brain - putamen (basal ganglia)              0.016789\n",
       "4         brain - substantia nigra              0.018051\n",
       "5                             skin              0.019730"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Tissue specific spoearman correlation plots\n",
    "alphagenome_spearman = variants.groupby('tissue').apply(\n",
    "    lambda x: x['slope'].corr(x['alphagenome_score'], method='spearman')\n",
    ").reset_index()\n",
    "alphagenome_spearman.columns = ['tissue', 'spearman_correlation']\n",
    "alphagenome_spearman"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a53fc24",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
